### PR TITLE
Parser Improvements and Additions

### PR DESCRIPTION
**Pull request by ShikharJ**

### CODE REVIEW

**Comment by ShikharJ** [comment]
This currently gives an error as:
```
/home/shikhar/symengine/symengine/parser.cpp:116:47: error: could not convert ‘{{"Eq", SymEngine::Eq}, {"Ne", SymEngine::Ne}, {"Ge", SymEngine::Ge}, {"Gt", SymEngine::Gt}, {"Le", SymEngine::Le}, {"Lt", SymEngine::Lt}}’ from ‘<brace-enclosed initializer list>’ to ‘std::map<std::__cxx11::basic_string<char>, std::function<Teuchos::RCP<const SymEngine::Boolean>(const Teuchos::RCP<const SymEngine::Basic>&, const Teuchos::RCP<const SymEngine::Basic>&)> >’
             {"Gt", Gt}, {"Le", Le}, {"Lt", Lt}};
```

**Comment by isuruf** [comment]
Looks good to me. It'll also be useful to be able to parse strings like `x < y` as well.

**Comment by ShikharJ** [comment]
@isuruf Can you review? I'm not sure this is how it is supposed to be implemented.

File: symengine/parser.cpp [hunk 1]
```diff
@@ -105,6 +108,15 @@ class ExpressionParser
             {"polygamma", polygamma},
             {"kronecker_delta", kronecker_delta}};
 
+    // maps string to corresponding double argument boolean function
+    std::map<std::string,
+             std::function<RCP<const Boolean>(const RCP<const Basic> &,
+                                              const RCP<const Basic> &)>>
+        double_arg_boolean_functions = {
```
Comment (srajangarg) [inline symengine/parser.cpp]: Can you rename this to `boolean_functions`?

File: symengine/parser.cpp [hunk 2]
```diff
@@ -119,7 +131,7 @@ class ExpressionParser
     // the string to be parsed, obtained after removing all spaces from input
     // string
     std::string s;
-    // it's length
+    // its length
```
Comment (srajangarg) [inline symengine/parser.cpp]: Grammar nazi! :stuck_out_tongue:

File: symengine/parser.cpp (hunk 1 above)
Comment (srajangarg) [inline symengine/parser.cpp]: The compilation error is possibly due to the overloaded `Eq` function. See how I have done it for the `log` function above.

File: symengine/parser.cpp [hunk 3]
```diff
@@ -196,6 +208,31 @@ class ExpressionParser
                                  parse_string(iter + 1, operator_end[iter]));
                     iter = operator_end[iter] - 1;
 
+                } else if (s[iter] == '=') {
+                    result = Eq(result,
+                                parse_string(iter + 1, operator_end[iter]));
+                    iter = operator_end[iter] - 1;
+
+                } else if (s[iter] == '<' and s[iter + 1] == '=') {
```
Comment (srajangarg) [inline symengine/parser.cpp]: can `iter` be the last index? `s[iter+1]` will segfault

**Comment by srajangarg** [comment]
I would like to see some more complicated cases, and cases which will not be parsed correctly using these new symbols

File: symengine/parser.cpp (hunk 3 above)
Comment (ShikharJ) [inline symengine/parser.cpp]: I have my doubts about that as well. A hacky alternative that occurred to me was to use other symbols, such as `#` or `@` for replacing all the instances of `<=` and `>=`, just like it's being done for `**` to `^` during preprocessing. What would you suggest?

Comment (srajangarg) [inline symengine/parser.cpp]: No that isn't a good solution. For now, will checking `iter + 1 < end` work?

Comment (ShikharJ) [inline symengine/parser.cpp]: It doesn't. Each time the terminal returns `Operator Inconsistency!`.

Comment (srajangarg) [inline symengine/parser.cpp]: Did you try and look into this further? Why does adding `&& iter + 1 < end` cause it to throw everytime?

Comment (ShikharJ) [inline symengine/parser.cpp]: Yes. Tried for `<` as well as `<=`.

Comment (ShikharJ) [inline symengine/parser.cpp]: @srajangarg Though I'm not sure, probably the error occurs in `parse_expr()` where `x >= y` get split into `x >` and the rest. I think it is during the simplification of this expression that the error is thrown up.

Comment (srajangarg) [inline symengine/parser.cpp]: So, ideally to fix this we should move from a character based approach to a "token" based approach, where each token is one or more characters. Everything proceeds the same way but instead of iterating over characters we iterate over tokens. Tokens are generated in the `parse_expr` stage. You can think of `**` being converted to `^` a tokenization itself (where we restrict tokens to be only characters). 

Do you think you can implement this? As we add more and more operators, using only single characters as tokens will become a problem (and we can soon run out of symbols). If you don't want to tackle this now, go ahead with the special symbol hacky approach.

**Comment by srajangarg** [comment]
what does `sin(x < y)` do?

**Comment by ShikharJ** [comment]
That gives out an `Operator Inconsistency` error as well.

File: symengine/parser.cpp (hunk 3 above)
Comment (ShikharJ) [inline symengine/parser.cpp]: I had tried changing the `set<char> OPERATORS` to `set<std::string> OPERATORS` and subsequently `std::map<char, int> op_precedence` to `std::map<std::string, int> op_precedence`. But these changes would require an overhaul of the current iterative algorithm. I'd like to open an issue, for now, and tackle it later.

**Comment by srajangarg** [comment]
Should this be an error while parsing (ie. you cant have this string), or error while actually constructing the symbolic tree (ie. a relational inside a `sin` should not be allowed)?

I feel it should parse correctly. I'm not sure.

**Comment by ShikharJ** [comment]
I think it should be handled in the respective classes. More recently, like it is handled in `Floor` and `Ceiling`. `SymPy` handles this in the classes as well.

**Comment by srajangarg** [comment]
So then our parser throwing `Operator Inconsistency` is wrong in the case of `sin(x < y)` right? Figure out what's going wrong, and try to fix it.

**Comment by ShikharJ** [comment]
@srajangarg Can you review? The build failure is unrelated to the changes, probably that needs to be restarted. `sin(x < y)` currently returns `sin(Lt(x, y))`, though that can be fixed through another PR.

**Comment by srajangarg** [comment]
Is the operator precedence set properly? How is `x + y < 2` parsed? Please add more extensive test cases, dealing with brackets, operators, functions etc.

**Comment by ShikharJ** [comment]
@srajangarg Can you restart the failing build? Also, should I add more tests?

File: symengine/tests/basic/test_parser.cpp [hunk 4]
```diff
@@ -259,6 +265,102 @@ TEST_CASE("Parsing: functions", "[parser]")
...
+
+    s = "x + y < x*y";
+    res = parse(s);
+    CHECK(eq(*res, *Lt(add(x, y), mul(x, y))));
+
+    s = "x + y >= x*y";
+    res = parse(s);
+    CHECK(eq(*res, *Le(mul(x, y), add(x, y))));
+
+    s = "x - y = x/y";
```
Comment (isuruf) [inline symengine/tests/basic/test_parser.cpp]: `=` should not be used for equality. Use `==`

File: symengine/tests/basic/test_parser.cpp [hunk 5]
```diff
@@ -259,6 +265,102 @@ TEST_CASE("Parsing: functions", "[parser]")
...
+    res = parse(s);
+    CHECK(eq(*res, *Eq(sub(x, y), div(x, y))));
+
+    s = "x - y <= x/y";
+    res = parse(s);
+    CHECK(eq(*res, *Le(sub(x, y), div(x, y))));
+
+    s = "x = y < 2";
```
Comment (isuruf) [inline symengine/tests/basic/test_parser.cpp]: This test and all the tests below should be removed. They don't make sense.

Comment (ShikharJ) [inline symengine/tests/basic/test_parser.cpp]: These were implemented to check for operator precedence. I'll remove them.

**Comment by ShikharJ** [comment]
Ping @isuruf @srajangarg

**Comment by srajangarg** [comment]
LGTM

File: symengine/tests/basic/test_parser.cpp [hunk 6]
```diff
@@ -269,6 +362,7 @@ TEST_CASE("Parsing: constants", "[parser]")
 
     s = "E*pi";
     res = parse(s);
+    s = "2*(x+1)**10 + 3*(x+2)**5";
```
Comment (isuruf) [inline symengine/tests/basic/test_parser.cpp]: Why this change?

Comment (ShikharJ) [inline symengine/tests/basic/test_parser.cpp]: Sorry, this is accidental.

**Comment by ShikharJ** [comment]
@isuruf @srajangarg I've added support for some additional `Boolean` functions. Please review the last commit.

**Comment by ShikharJ** [comment]
Ping @isuruf @srajangarg.

File: symengine/parser.cpp [hunk 7]
```diff
@@ -374,11 +485,27 @@ class ExpressionParser
         s.clear();
         s.reserve(in.length());
 
-        // Replacing ** with ^
+        // TODO: Implement multi-character operator parsing support
```
Comment (isuruf) [inline symengine/parser.cpp]: I don't really like this hack. Would it take a long time to implement this?

File: symengine/parser.cpp [hunk 8]
```diff
@@ -370,13 +503,13 @@ class ExpressionParser
...
-        char last_char = 'x';
+        std::string last_char = "x";
         s.clear();
         s.reserve(in.length());
 
-        // Replacing ** with ^
         for (unsigned int i = 0; i < in.length(); ++i) {
             if (in[i] == '*' and i + 1 < in.length() and in[i + 1] == '*') {
+                // Replacing ** with ^
                 s += '^';
```
Comment (ShikharJ) [inline symengine/parser.cpp]: @isuruf @srajangarg Should this be removed? Is there a need to parse `&`, `|` or `^`?

**Comment by ShikharJ** [comment]
Ping @isuruf.

File: symengine/parser.cpp (hunk 8 above)
Comment (srajangarg) [inline symengine/parser.cpp]: Yes it should. If you've implemented multi operator support.

**Comment by srajangarg** [comment]
first look this looks good. give me some time to go through it fully.

**Comment by ShikharJ** [comment]
@isuruf What would be your take on this? I don't have a clear idea on implementing "tokenization" of operators, and as such, I'd like to open an issue for that instead.
